lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -482,10 +482,105 @@ struct WarpOpTransferWrite : public OpRewritePattern<vector::TransferWriteOp> {
+///     ...
+///     scf.yield %r : vector<128xf32>
+///   }
+///   vector_ext.yield %v1 : vector<128xf32>
+/// }
+/// ```
+/// To:
+/// %w0 = vector_ext.warp_execute_on_lane_0(%arg0) -> (vector<4xf32>) {
+///   ...
+///   vector_ext.yield %v : vector<128xf32>
+/// }
+/// %w = scf.for %arg3 = %c0 to %c128 step %c1 iter_args(%varg = %q0)
+///   -> (vector<4xf32>) {
+///     %iw = vector_ext.warp_execute_on_lane_0(%laneid)
+///     args(%varg : vector<4xf32>) -> (vector<4xf32>) {
+///     ^bb0(%arg: vector<128xf32>):
+///       ...
+///       vector_ext.yield %ir : vector<128xf32>
+///     }
+///     scf.yield %iw : vector<4xf32>
+///  }
+/// ```
+struct WarpOpScfForOp : public OpRewritePattern<WarpSingleLaneOp> {
+  using OpRewritePattern<WarpSingleLaneOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(WarpSingleLaneOp warpOp,
+                                PatternRewriter &rewriter) const override {
+    auto yield = cast<vector_ext::YieldOp>(
+        warpOp.getBodyRegion().getBlocks().begin()->getTerminator());
+    // Only pick up forOp if it is the last op in the region.
+    Operation *lastNode = yield->getPrevNode();
+    auto forOp = dyn_cast_or_null<scf::ForOp>(lastNode);
+    if (!forOp)
+      return failure();
+    SmallVector<Value> newOperands;
+    SmallVector<unsigned> resultIdx;
+    // Collect all the outputs coming from the forOp.
+    for (OpOperand &yieldOperand : yield->getOpOperands()) {
+      if (yieldOperand.get().getDefiningOp() != forOp.getOperation())
+        continue;
+      newOperands.push_back(warpOp.getResult(yieldOperand.getOperandNumber()));
+      yieldOperand.set(
+          forOp.getIterOperands()[yieldOperand.getOperandNumber()]);
+      resultIdx.push_back(yieldOperand.getOperandNumber());
+    }
+    OpBuilder::InsertionGuard g(rewriter);
+    rewriter.setInsertionPointAfter(warpOp);
+    // Create a new for op outside the region with a WarpSingleOp region inside.
+    auto newForOp = rewriter.create<scf::ForOp>(
+        forOp.getLoc(), forOp.getLowerBound(), forOp.getUpperBound(),
+        forOp.getStep(), newOperands);
+    rewriter.setInsertionPoint(newForOp.getBody(), newForOp.getBody()->begin());
+    auto innerWarp = rewriter.create<WarpSingleLaneOp>(
+        warpOp.getLoc(), newForOp.getResultTypes(), warpOp.laneid(),
+        newForOp.getRegionIterArgs(), forOp.getResultTypes());
+    // Move the loop region within the new WarpSingleLaneOp region.
+    BlockAndValueMapping mapping;
+    mapping.map(forOp.getInductionVar(), newForOp.getInductionVar());
+    for (auto args : llvm::zip(forOp.getRegionIterArgs(),
+                               innerWarp.getBody()->getArguments())) {
+      mapping.map(std::get<0>(args), std::get<1>(args));
+    }
+    rewriter.setInsertionPoint(innerWarp.getBody(),
+                               innerWarp.getBody()->begin());
+    for (Operation &innerOp : forOp.getBody()->without_terminator())
+      rewriter.clone(innerOp, mapping);
```

Comment:
I would use `RewriterBase::mergeBlocks`. You can move over the entire loop body to a new loop. See `Dialect/SCF/Transforms/BufferizableOpInterfaceImpl.cpp:370` for an example.

Reply:
That's what I add tried originally but the blocks don't exactly match as the scf.for as the induction variable as extra block argument.

Reply:
I think it should still work. You can use `argValues` to pass replacement values. In the same way you are using the BlockAndValueMapping at the moment.

Reply:
I think an scf::ForOp block always starts with the IV, followed by all iter_args. So you should be able to pass `/*argValues=*/{newForOp.getInductionVar(), innerWarp.getBody()->getArguments() /*expanded*/}`.

Actually, it's not going to be one-liner as I hoped... So not sure if it's worth it.